### DISCUSSION TITLE
Rules how Roslyn generates IL code

### DISCUSSION THREAD

**Discussion by ViktoriaNiko** [dotnet/roslyn#55567] [General]
We work on .NET tool for real-time application profilin. It provides the ability to make white box functional testing.
We work with IL code and need to calculate coverage. To calculate it correctly we need to understand which parts of IL are Business code and which are not. For example, for async\await we don’t need to consider such instructions as some internal checks, starting of async state machine and so on. Async\await of course not the only thing that looks quite complicated in IL. Unfortunately, we can’t find any documentation for it.
It there any documentation how Roslyn generates IL code for different statements, methods and so on? Which rules does it use?

**Comment by svick** [comment]
I *think* one of the big reasons this is undocumented is because having documentation effectively says "you can rely on this", and so it makes it harder to change things in the future. But, since Roslyn is open source, you can always look at the source code. (Though I understand that navigating a large unfamiliar codebase is a fairly high barrier to overcome.)

Though, instead of relying on what IL the compiler produces today, have you considered using debugging information? You can use that to map ranges of IL instructions back to pieces of C# code, which could give you what you're looking for.
